### PR TITLE
Add context

### PR DESCRIPTION
**Pull request by osterman**
## what
* Support passing a label's context between label modules

## why
* DRY

## demo

```
module "label1" {
  source     = "../../"
  namespace  = "Namespace"
  stage      = "Stage"
  name       = "Name1"
  attributes = ["1", "2", "3", ""]
  tags       = {
    "SomeKey" = "SomeValue"
  }
}

module "label2" {
  source     = "../../"
  context    = "${module.label1.context}"
  name       = "Name2"
}
```

## todo
- [x] figure out how to inherit extra tags from `var.tags` <--- need help

### CODE REVIEW

**Comment by osterman** [comment]
This is working and produces the following output. I just don't know how to inherit the additional tags from the parent's context. 

```
Outputs:

label1_attributes = 1-2-3
label1_context = {
  attributes = 1-2-3
  delimiter = -
  name = name
  namespace = namespace
  stage = stage
  tags = map[Namespace:namespace SomeKey:SomeValue Stage:stage Name:namespace-stage-name-1-2-3]
}
label1_id = namespace-stage-name-1-2-3
label1_name = name
label1_namespace = namespace
label1_stage = stage
label1_tags = {
  Name = namespace-stage-name-1-2-3
  Namespace = namespace
  SomeKey = SomeValue
  Stage = stage
}
label2_attributes = 1-2-3
label2_context = {
  attributes = 1-2-3
  delimiter = -
  name = test
  namespace = namespace
  stage = stage
  tags = map[Namespace:namespace Stage:stage Name:namespace-stage-test-1-2-3]
}
label2_id = namespace-stage-test-1-2-3
label2_name = test
label2_namespace = namespace
label2_stage = stage
label2_tags = {
  Name = namespace-stage-test-1-2-3
  Namespace = namespace
  Stage = stage
}
```

File: main.tf [hunk 1]
```diff
@@ -1,22 +1,36 @@
...
+  stage      = "${var.stage == "" ? lookup(var.context, "stage", "") : lower(format("%v", var.stage))}"
+  attributes = "${length(var.attributes) == 0 ? lookup(var.context, "attributes", "") : lower(format("%v", join(var.delimiter, compact(var.attributes))))}"
+
+  generated_tags = {
+    "Name"      = "${local.enabled ? local.id : ""}"
+    "Namespace" = "${local.enabled ? local.namespace : ""}"
+    "Stage"     = "${local.enabled ? local.stage : ""}"
+  }
+
+  tags = "${merge(local.generated_tags, var.tags)}"
```
Comment (osterman) [inline main.tf]: Need to incorporate the `var.context["tags"]` here some how

**Comment by dennybaa** [comment]
@osterman Hey, it's an awesome idea to have context.

Unfortunately I also don't see how it's possible with HCL1. We should always try to remember that scalars such as maps/list have always to be flat. Basically scalars must:
- be homogeneous
- include only simple types as string/integer/boolean

These limitations will be hopefully resolved in 0.12 if it includes HCL2...

In other words now all the variables and all the outputs should be flat and homogeneous. With the introduced context you break this rule.

Regarding non-flat scalars, the only workaround possible now is leveraging **locals** for intermediate computation, like bellow:

```hcl

locals {
  non_flat_map = {
    map1 = "${var.map1}"  # var.map1 should be flat
    map2 = "${var.map2}"  # var.map2 should be flat
  }

  target_map = "${local.non_flat_map[var.choice ? "map1": "map2"]}"
}

output "map" {
  value = "${local.target_map}"
}
```

File: main.tf (hunk 1 above)
Comment (dennybaa) [reply main.tf]: it doesn't seem possible :(

**Comment by Jamie-BitFlight** [comment]
### This code:
```hcl
module "label1" {
  source     = "../../"
  namespace  = "CloudPosse"
  stage      = "Production"
  name       = "Winston"
  attributes = ["fire", "water", "earth", "air"]

  tags = {
    "City"        = "Dublin"
    "Environment" = "Private"
  }
}

module "label2" {
  source  = "../../"
  context = "${module.label1.context}"
  name    = "Charlie"

  tags = {
    "City"        = "London"
    "Environment" = "Public"
  }
}
output "label1" {
  value = {
    id         = "${module.label1.id}"
    name       = "${module.label1.name}"
    namespace  = "${module.label1.namespace}"
    stage      = "${module.label1.stage}"
    attributes = "${module.label1.attributes}"
  }
}

output "label2" {
  value = {
    id         = "${module.label2.id}"
    name       = "${module.label2.name}"
    namespace  = "${module.label2.namespace}"
    stage      = "${module.label2.stage}"
    attributes = "${module.label2.attributes}"
  }
}

output "label1_tags" {
  value = "${module.label1.tags}"
}

output "label1_context" {
  value = "${module.label1.context}"
}

output "label2_tags" {
  value = "${module.label2.tags}"
}

output "label2_context" {
  value = "${module.label2.context}"
}
```
### Makes this output now
```
label1 = {
  attributes = [fire water earth air]
  id = cloudposse-production-winston-fire-water-earth-air
  name = winston
  namespace = cloudposse
  stage = production
}
label1_context = {
  attributes = [fire water earth air]
  delimiter = [-]
  name = [winston]
  namespace = [cloudposse]
  stage = [production]
  tags_keys = [City Environment Name Namespace Stage]
  tags_values = [Dublin Private cloudposse-production-winston-fire-water-earth-air cloudposse production]
}
label1_tags = {
  City = Dublin
  Environment = Private
  Name = cloudposse-production-winston-fire-water-earth-air
  Namespace = cloudposse
  Stage = production
}
label2 = {
  attributes = [fire water earth air]
  id = cloudposse-production-winstoncharlie-fire-water-earth-air
  name = winstoncharlie
  namespace = cloudposse
  stage = production
}
label2_context = {
  attributes = [fire water earth air]
  delimiter = [-]
  name = [winstoncharlie]
  namespace = [cloudposse]
  stage = [production]
  tags_keys = [City Environment Name Namespace Stage]
  tags_values = [London Public cloudposse-production-winston-fire-water-earth-air cloudposse production]
}
label2_tags = {
  City = London
  Environment = Public
  Name = cloudposse-production-winston-fire-water-earth-air
  Namespace = cloudposse
  Stage = production
}
```

**Comment by Jamie-BitFlight** [comment]
@dennybaa I got it working

File: main.tf [hunk 2]
```diff
@@ -1,26 +1,53 @@
...
+  selected_stage      = ["${compact(concat(local.context_local["stage"], list(var.stage)))}"]
+  stage               = "${lower(join("", split(" ", local.selected_stage[0])))}"
+  selected_attributes = ["${distinct(compact(concat(var.attributes, local.context_local["attributes"])))}"]
+  attributes          = "${split("~^~", lower(join("~^~", local.selected_attributes)))}"
+  selected_delimiter  = ["${distinct(compact(concat(local.context_local["delimiter"], list(var.delimiter))))}"]
+  delimiter           = "${join("", split(" ", local.selected_delimiter[0]))}"
+  # Merge the map of empty values, with the variable context, so that context_local always contains all map keys
+  context_local = "${merge(local.context_context, var.context)}"
+  # Only maps that contain all the same attribute types can be merged, so they have been set to list
+  context_context = {
```
Comment (osterman) [inline main.tf]: can we rename this? maybe call it `null_context`?

File: main.tf [hunk 3]
```diff
@@ -1,26 +1,53 @@
...
+  # name: Remove spaces, make lowercase
+
+  selected_name       = ["${compact(concat(list(var.name), local.context_local["name"]))}"]
+  name                = "${lower(join("", split(" ", local.selected_name[0])))}"
+  selected_namespace  = ["${compact(concat(local.context_local["namespace"], list(var.namespace)))}"]
+  namespace           = "${lower(join("", split(" ", local.selected_namespace[0])))}"
+  selected_stage      = ["${compact(concat(local.context_local["stage"], list(var.stage)))}"]
+  stage               = "${lower(join("", split(" ", local.selected_stage[0])))}"
+  selected_attributes = ["${distinct(compact(concat(var.attributes, local.context_local["attributes"])))}"]
+  attributes          = "${split("~^~", lower(join("~^~", local.selected_attributes)))}"
```
Comment (osterman) [inline main.tf]: let's call `~^~` something (E.g. `unique_separator`) - either a `variable` or a `local`.

File: main.tf (hunk 2 above)
Comment (Jamie-BitFlight) [reply main.tf]: I'm renaming it to `context_struct`, since it is the structure of context with all variables.

**Comment by osterman** [comment]
@dennybaa - please have another look at this. It's not pretty, but it sets out to accomplish what we wanted to do. I'm pretty sure we'll end up merging this.

**Comment by dennybaa** [comment]
@osterman Wow. That's a lot of effort, this might be a good example of that how to mitigate passing non-flat structures :) 

It would be nice if you also switch to **null_data_source** this reduces output rubbish during plan/apply, and specifically made for this, in other words resource is an overkill here.

This would work now, though tags passing seemed _reasonably_ easier and if I'm not mistaken achieved the same goal. My personal opinion that terraform is not yet ready for things we are trying to do... Don't get me wrong but it doesn't feel like an easy way. Though I really admire your approach and implementation :+1:

**Comment by dennybaa** [comment]
So I still think, taking an approach of calling tags Name/Namespace/Stage as generated_tags. IMO it still looks neater for me to have an option like `exclude_generated_tags`, because it will generally keep locals computations simpler.

**Comment by Jamie-BitFlight** [comment]
I think that this will get far easier in tf 0.12

**Comment by dennybaa** [comment]
@osterman yeah, it looks a bit complicated, but what we can do :) The module is very usable now :+1: for it.

**Comment by dennybaa** [comment]
@osterman Hey, Erik. I've noticed it seems that you missed https://github.com/stackfeed/terraform-null-label/commit/4d7fb86ab793edb4664ab7bef422fc80b84bfcd0

**Pull request merged**